### PR TITLE
Fix "go to latest version" for /src/ when an item was renamed

### PR DESCRIPTION
**Pull request by jyn514**
Fixes https://github.com/rust-lang/docs.rs/issues/513.

### CODE REVIEW

File: src/test/fakes.rs [hunk 1]
```diff
@@ -251,8 +251,16 @@ impl<'a> FakeRelease<'a> {
...
+                let mut updated = ["src", &package.name, src].join("/");
+                updated += ".html";
+                let source_html = format!(
+                    "<html><head></head><body>{}</body></html>",
+                    std::str::from_utf8(data).expect("invalid utf8")
+                );
+                rustdoc_files.push((
+                    Box::leak(Box::new(updated)),
+                    Box::leak(source_html.into_bytes().into_boxed_slice()),
+                ));
```
Comment (jyn514) [inline src/test/fakes.rs:263]: This took me *so* long to debug - the HTML rewriter doesn't actually do anything unless at least these tags exist.

File: src/test/fakes.rs [hunk 2]
```diff
@@ -264,9 +272,14 @@ impl<'a> FakeRelease<'a> {
             fs::create_dir(&path_prefix)?;
 
             for (path, data) in files {
+                if path.starts_with("/") {
+                    failure::bail!("absolute paths not supported");
+                }
```
Comment (jyn514) [inline src/test/fakes.rs:277]: This also took ages to debug.

**Comment by jyn514** [comment]
@syphar do you have time to review this?

**Comment by syphar** [comment]
@jyn514 I will check it out, probably this weekend.

File: src/web/rustdoc.rs [hunk 3]
```diff
@@ -493,6 +493,11 @@ fn path_for_version(
     } else {
         ""
     };
+    let is_source_view = if platform.is_empty() {
+        req_path.get(3).copied() == Some("src")
```
Comment (syphar) [inline src/web/rustdoc.rs:497]: without knowing all the connections too well, quite some digging is necessary to know which two cases we are handling here. 

Could we add a comment here?

**Review by syphar — CHANGES_REQUESTED** [review]
So, I went through it, as far as I know the code. 

In general I remember that especially all this redirecting was touched to often that it's actually covered with tests quite well. (please correct me if I'm wrong). 

Apart from my wish for an explaining comment, there is only one thing that I found in manual testing: 

When the old version's URL had a line-number attached (with `#100` for example), it would also end up in the search-URL. Since it does _not_ end up in the search-field, we could just leave it.

**Comment by jyn514** [comment]
> In general I remember that especially all this redirecting was touched to often that it's actually covered with tests quite well. (please correct me if I'm wrong).

Yes, redirects are tested quite well (I think they may have been the first thing we added tests for).

> When the old version's URL had a line-number attached (with #100 for example), it would also end up in the search-URL. Since it does not end up in the search-field, we could just leave it.

Hmm, I don't mind keeping it (working on that now), but it shouldn't make a difference one way or another like you said.

File: src/web/rustdoc.rs (hunk 3 above)
Comment (jyn514) [reply src/web/rustdoc.rs:497]: Sure, I left a couple comments - let me know if they make sense.

**Review by syphar — APPROVED** [review]

**Comment by jyn514** [comment]
> When the old version's URL had a line-number attached (with #100 for example), it would also end up in the search-URL. 

I just realized this is a pre-existing bug - I'd like to leave it for a follow up. Feel free to open an issue for it :)

**Pull request merged**